packages/vscode-extension/src/debugging/DebugSession.ts

```diff
@@ -22,134 +22,148 @@ export type DebugSessionDelegate = {
-        didStartHandler?.dispose();
-        didStartHandler = null;
-      }
-    });
-    try {
-      const debugStarted = await debug.startDebugging(
-        undefined,
-        {
-          type: REACT_NATIVE_SESSION_TYPE,
-          name: "React Native Preview Debugger",
-          request: "attach",
-        },
-        {
-          suppressDebugStatusbar: true,
-          suppressDebugView: true,
-          suppressDebugToolbar: true,
-          suppressSaveBeforeStart: true,
-        }
-      );
-
-      if (debugStarted) {
-        // NOTE: this is safe, because `debugStarted` means the session started successfully,
-        // and we set the session in the `onDidStartDebugSession` handler
-        assert(this.vscSession, "Expected debug session to be set");
-        return true;
+  public async startParentDebugSession() {
+    this.parentDebugSession = await startDebugging(
+      undefined,
+      {
+        type: MASTER_DEBUGGER_TYPE,
+        name: "Radon IDE Debugger",
+        request: "attach",
+      },
+      {
+        suppressDebugStatusbar: true,
+        suppressDebugView: true,
+        suppressDebugToolbar: true,
+        suppressSaveBeforeStart: true,
       }
-    } finally {
-      didStartHandler?.dispose();
-    }
+    );
   }
 
   public static start(debugEventDelegate: DebugSessionDelegate) {
-    const debugSession = new DebugSession(debugEventDelegate);
-    debugSession.startInternal();
+    const debugSession = new DebugSession(true, debugEventDelegate);
+    debugSession.startParentDebugSession();
     return debugSession;
   }
 
-  public async getOriginalSource(
-    fileName: string,
-    line0Based: number,
-    column0Based: number
-  ): Promise<{ sourceURL: string; lineNumber1Based: number; columnNumber0Based: number }> {
-    return await this.session.customRequest("source", { fileName, line0Based, column0Based });
-  }
-
   public async restart() {
     await this.stop();
-    await this.startInternal();
+    if (this.useParentDebugSession) {
+      await this.startParentDebugSession();
+    }
   }
 
   private async stop() {
-    this.vscSession && (await debug.stopDebugging(this.vscSession));
-    this.vscSession = undefined;
-    this.wasConnectedToCDP = false;
+    if (this.parentDebugSession) {
+      await debug.stopDebugging(this.parentDebugSession);
+    }
+    if (this.jsDebugSession) {
+      await debug.stopDebugging(this.jsDebugSession);
+    }
+    this.parentDebugSession = undefined;
+    this.jsDebugSession = undefined;
     this.currentWsTarget = undefined;
   }
 
-  /**
-  This method is async to allow for awaiting it during restarts, please keep in mind tho that
-  build in vscode dispose system ignores async keyword and works synchronously.
-  */
-  public async dispose() {
-    this.vscSession && (await debug.stopDebugging(this.vscSession));
-    this.debugEventsListener.dispose();
+  public dispose() {
+    this.stop();
+    disposeAll(this.disposables);
```

Comment:
NIT: Does the order here matter?
In particular, the way it is written now, it's not clear to me if `delegate.onDebugSessionTerminated` here will be called:
- on one hand, `stop` will terminate the session, and we call it before disposing the `terminate` event subscription.
- on the other, `stop` is async, so I'd assume the actual termination will happen in the future, after the subscription is already disposed.

I think I'd prefer if we either called `disposeAll` first, or after the promise from `stop` resolved.

```suggestion
    this.stop().then(() => disposeAll(this.disposables));
```